services/web/server/tests/unit/with_dbs/03/invitations/test_products__invitations_handlers.py

```diff
@@ -104,3 +105,105 @@ async def test_product_owner_generates_invitation(
     assert before_dt < got.created
     assert got.created < datetime.now(tz=timezone.utc)
+
+
+MANY_TIMES: Final = 2
+
+
+@pytest.mark.acceptance_test(
+    "pre-registration in https://github.com/ITISFoundation/osparc-simcore/issues/5138"
+)
+@pytest.mark.parametrize(
+    "user_role,expected_status",
+    [
+        (UserRole.PRODUCT_OWNER, status.HTTP_200_OK),
+    ],
+)
+async def test_pre_registration_and_invitation_workflow(
+    client: TestClient,
+    mock_invitations_service_http_api: AioResponsesMock,
+    logged_user: UserInfoDict,
+    expected_status: HTTPStatus,
+    guest_email: str,
+    user_password: str,
+    faker: Faker,
+):
+    requester_info = {
+        "firstName": faker.first_name(),
+        "lastName": faker.last_name(),
+        "email": guest_email,
+        "companyName": faker.company(),
+        "phone": faker.phone_number(),
+        # billing info
+        "address": faker.address().replace("\n", ", "),
+        "city": faker.city(),
+        "state": faker.state(),
+        "postalCode": faker.postcode(),
+        "country": faker.country(),
+    }
+
+    invitation = GenerateInvitation(
+        guest=guest_email,
+        trial_account_days=None,
+        extra_credits_in_usd=10,
+    ).dict()
+
+    # Search user -> nothing
+    response = await client.get("/v0/users:search", params={"email": guest_email})
+    data, _ = await assert_status(response, expected_status)
+    # i.e. no info of requester is found, i.e. needs pre-registration
+    assert data == []
+
+    # Cannot generate anymore an invitation for users that are not registered or pre-registered.
+    # Cannot do that because otherwise old users willing a new product will not work!
+    # response = await client.post("/v0/invitation:generate", json=invitation)
+    # assert response.status == status.HTTP_409_CONFLICT
+
+    # Accept user for registration and create invitation for her
+    response = await client.post("/v0/users:pre-register", json=requester_info)
+    data, _ = await assert_status(response, expected_status)
+
+    # Can only  pre-register once
+    for _ in range(MANY_TIMES):
+        response = await client.post("/v0/users:pre-register", json=requester_info)
+        await assert_status(response, status.HTTP_409_CONFLICT)
+
+    # Search user again
+    for _ in range(MANY_TIMES):
+        response = await client.get("/v0/users:search", params={"email": guest_email})
+        data, _ = await assert_status(response, expected_status)
+        assert len(data) == 1
+        user_found = data[0]
+        assert not user_found["registered"]
+        assert user_found["email"] == guest_email
+
+    # Can make as many invitations as I wish
+    for _ in range(MANY_TIMES):
+        response = await client.post("/v0/invitation:generate", json=invitation)
+        data, _ = await assert_status(response, status.HTTP_200_OK)
+        assert data["guest"] == guest_email
+        got_invitation = InvitationGenerated.parse_obj(data)
```

Comment:
Q: would the generated invitation always be the same, given the same data?

Reply:
yes, provided that it uses the same secret to encrypt